kernel/proc.c

```diff
@@ -169,6 +172,42 @@ freeproc(struct proc *p)
   p->killed = 0;
   p->xstate = 0;
   p->state = UNUSED;
+  p->tickets = 0;
+  p->pass = 0;
+  p->ticks = 0;
+}
+
+// 0으로 초기화되어도 `get_runnable_min_pass_proc`에서 적절한 값으로 덮어씌워진다.
+// 프로그램 초기에 덮어씌워지지 않은 값이 읽혀서 사용되더라도 프로그램 초기에는
+// 실제 최소 pass 값이 0이므로 큰 이슈가 없을 것으로 예상?
+// _Atomic을 사용하여 멀티코어 환경에서 race condition 방지
+static _Atomic uint64 cached_min_pass = 0;
+
+// 주의) proc에 대한 락을 잡은 상태에서 호출하면 중복 acquire로 패닉 발생할 수 있음
+static struct proc*
+get_runnable_min_pass_proc_locked(void) {
+  uint64 min_pass = ((uint64)~0ULL); // UINT64_MAX
+  struct proc* min_pass_proc = 0;
+  for(int i = 0; i < NPROC; i++) {
+    acquire(&proc[i].lock);
+    if(proc[i].state == RUNNABLE) {
+      if(proc[i].pass < min_pass) {
+        // 이전 min_pass_proc이 존재한다면 락을 해제  
+        if(min_pass_proc != 0) {
+          release(&min_pass_proc->lock);
+        }
+        min_pass = proc[i].pass;
+        min_pass_proc = &proc[i];
+        // 새 min_pass_proc에 대한 락은 유지
+      } else {
+        release(&proc[i].lock);
+      }
+    } else {
+      release(&proc[i].lock);
+    }
+  }
```

Comment:
요렇게 짜면 로직이 좀 더 단순합니다

```suggestion
  for(int i = 0; i < NPROC; i++) {
    acquire(&proc[i].lock);

    if(proc[i].state == RUNNABLE && proc[i].pass < min_pass) {
      // 이전 min_pass_proc이 존재한다면 락을 해제
      if(min_pass_proc != 0) { release(&min_pass_proc->lock); }

      min_pass = proc[i].pass;
      min_pass_proc = &proc[i];

      // 새 min_pass_proc에 대한 락을 유지한채로, 다음 프로세스 검사
      continue;
    }

    release(&proc[i].lock);
  }
```